src/library/trackcollectionmanager.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <QDir>
+#include <QList>
+#include <QObject>
+#include <QSet>
+
+#include "library/scanner/libraryscanner.h"
+#include "preferences/usersettings.h"
+#include "track/globaltrackcache.h"
+#include "util/db/dbconnectionpool.h"
+
+class TrackCollection;
+class ExternalTrackCollection;
+
+// Manages the internal and a varying number of external track collections.
+//
+// All modifying operations that might affect external collections
+// must be invoked through this class to keep all track collections
+// synchronized!
+//
+// Both crates and playlists are currently only supported by the internal
+// collection, which needs to be modified directly.
+class TrackCollectionManager: public QObject,
+    public virtual /*implements*/ GlobalTrackCacheSaver {
+    Q_OBJECT
+
+  public:
+    TrackCollectionManager(
+            QObject* parent,
+            UserSettingsPointer pConfig,
+            mixxx::DbConnectionPoolPtr pDbConnectionPool,
+            deleteTrackFn_t deleteTrackFn = nullptr);
+    ~TrackCollectionManager() override;
+
+    TrackCollection* internalCollection() {
+        return m_pInternalCollection;
+    }
+
+    const QList<ExternalTrackCollection*>& externalCollections() const {
+        return m_externalCollections;
+    }
+
+    bool hideTracks(const QList<TrackId>& trackIds);
+    bool unhideTracks(const QList<TrackId>& trackIds);
+    void hideAllTracks(const QDir& rootDir);
+
+    void purgeTracks(const QList<TrackId>& trackIds);
+    void purgeAllTracks(const QDir& rootDir);
+
+    bool addDirectory(const QString& dir);
+    bool removeDirectory(const QString& dir);
+    void relocateDirectory(QString oldDir, QString newDir);
+
+    TrackPointer getOrAddTrack(
+            const TrackRef& trackRef,
+            bool* pAlreadyInLibrary = nullptr);
+
+    // Save the track in both the internal database and external collections.
+    // Export of metadata is deferred until the track is evicted from the
+    // cache to prevent file corruption due to concurrent access.
+    // Returns true if the track was dirty and has been saved, otherwise
+    // false.
+    bool saveTrack(const TrackPointer& pTrack);
+
+  signals:
+    void libraryScanStarted();
+    void libraryScanFinished();
+
+  public slots:
+    void startLibraryScan();
+    void stopLibraryScan();
+
+    void slotScanTrackAdded(TrackPointer pTrack);
+    void slotScanTracksUpdated(QSet<TrackId> updatedTrackIds);
+    void slotScanTracksReplaced(QList<QPair<TrackRef, TrackRef>> replacedTracks);
+
+  private:
+    // Callback for GlobalTrackCache
+    void saveEvictedTrack(Track* pTrack) noexcept override;
+
+    // Might be called from any thread
+    enum class TrackMetadataExportMode {
+        Immediate,
+        Deferred,
+    };
+    void saveTrack(
+            Track* pTrack,
+            TrackMetadataExportMode mode);
```

Comment:
This seems a bit odd to me to pass a TrackMetadataExportMode object to this function and `exportTrackMetadata` below. How about making a private `m_trackMetadataExportMode` member of TrackCollectionManager instead?

Reply:
Oh, I think I misunderstood. TrackMetadataExportMode isn't a user preference option, is it?

Reply:
This parameter lies the ground for the new, more robust handling of metadata synchronization between the database and file tags that will follow. The function `TrackCollectionManager::saveTrack()` which uses `Deferred` is still unused. Saving to the database is allowed at any time while saving to file tags is deferred until the last reference to the track goes out of scope.

Reply:
Okay